From: Kemeng Shi <shikemeng@huaweicloud.com>
Date: Thu, 10 Apr 2025 23:39:08 +0800
Subject: [PATCH] mm: swap: add __maybe_unused attribute for swap_is_last_ref()
 and update it's comment

Add __maybe_unused attribute for swap_is_last_ref ()to fix following build
warning:

mm/swapfile.c:1517:20: warning: function 'swap_is_last_ref' is not needed and will not be emitted [-Wunneeded-internal-declaration]
    1517 | static inline bool swap_is_last_ref(unsigned char count)
         |                    ^~~~~~~~~~~~~~~~
   1 warning generated.

Besides, original comment for swap_entries_free() is placed before
swap_is_last_ref() which may introduce confusion.  Update comment to
address this.

Link: https://lkml.kernel.org/r/20250410153908.612984-1-shikemeng@huaweicloud.com
Fixes: 6bb001b6b64e ("mm: swap: enable swap_entry_range_free() to drop any kind of last ref")
Signed-off-by: Kemeng Shi <shikemeng@huaweicloud.com>
Suggested-by: Baoquan He <bhe@redhat.com>
Reviewed-by: Baoquan He <bhe@redhat.com>
Tested-by: SeongJae Park <sj@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---
 mm/swapfile.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/mm/swapfile.c b/mm/swapfile.c
index 59e22f0b46c6..c6b4c74622fc 100644
--- a/mm/swapfile.c
+++ b/mm/swapfile.c
@@ -1520,15 +1520,19 @@ static bool swap_entries_put_nr(struct swap_info_struct *si,
 }
 
 /*
- * Drop the last ref(1, SWAP_HAS_CACHE or SWAP_MAP_SHMEM) of swap entries,
- * caller have to ensure all entries belong to the same cgroup and cluster.
+ * Check if it's the last ref of swap entry in the freeing path.
+ * Qualified vlaue includes 1, SWAP_HAS_CACHE or SWAP_MAP_SHMEM.
  */
-static inline bool swap_is_last_ref(unsigned char count)
+static inline bool __maybe_unused swap_is_last_ref(unsigned char count)
 {
 	return (count == SWAP_HAS_CACHE) || (count == 1) ||
 	       (count == SWAP_MAP_SHMEM);
 }
 
+/*
+ * Drop the last ref of swap entries, caller have to ensure all entries
+ * belong to the same cgroup and cluster.
+ */
 static void swap_entries_free(struct swap_info_struct *si,
 			      struct swap_cluster_info *ci,
 			      swp_entry_t entry, unsigned int nr_pages)
-- 
2.39.5

